Add Builder API (#209)
ethereum/execution-apis: src/builder/specification.md

    Note over relays: validate signature
    relays-->>mev_boost: builder_getPayload response
    Note over mev_boost: verify response matches expected
    mev_boost-->>consensus: builder_getPayload response
```

## Structures

### `ExecutionPayloadV1`

Mirror of [`ExecutionPayloadV1`][execution-payload].

### `ExecutionPayloadHeaderV1`

Equivalent to `ExecutionPayloadV1`, except `transactions` is replaced with `transactionsRoot`.
- `parentHash`: `DATA`, 32 Bytes
- `feeRecipient`:  `DATA`, 20 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactionsRoot`: `DATA`, 32 Bytes

#### SSZ Objects

Consider the following definitions supplementary to the definitions in [`consensus-specs`][consensus-specs].

##### `builder_setFeeRecipientV1` Request

```python
class SetFeeRecipientRequestV1(Container):
    feeRecipient: Bytes20

>     timestamp: uint64

Author: I'm less attached to timestamp over sequence number now. @ralexstokes has a compelling argument for switching: https://github.com/flashbots/mev-boost/issues/97

    deposits: List[Deposit, MAX_DEPOSITS]
    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
    sync_aggregate: SyncAggregate
    execution_payload: ExecutionPayloadHeader
```

## Errors

The list of error codes introduced by this specification can be found below.

| Code | Message | Meaning |
| - | - | - |
| -32000 | Server error | Generic client error while processing request. |
| -32001 | Unknown hash | No block with the provided hash is known. |
| -32002 | Unknown validator | No known mapping between validator and feeRecipient. |
| -32003 | Invalid SSZ | Unable to decode SSZ. |
| -32004 | Unknown block | Block does not match the provided header. |
| -32005 | Invalid signature | Provided signature is invalid. |
| -32006 | Invalid timestamp | Provided timestamp was invalid. |
| -32600 | Invalid request | The JSON sent is not a valid Request object. |
| -32601 | Method not found | The method does not exist / is not available. |
| -32602 | Invalid params | Invalid method parameter(s). |
| -32603 | Internal error | Internal JSON-RPC error. |
| -32700 | Parse error | Invalid JSON was received by the server. |

## Routines

### Signing

> All signature operations should follow the [standard BLS operations][bls] interface defined in `consensus-specs`.

Reviewer: ```suggestion
All signature operations should follow the [standard BLS operations][https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#bls-signatures] interface defined in `consensus-specs`.
```

Author: This is linked at the bottom using markdown reference style.

There are two types of data to sign over in the Builder API:
* In-protocol messages, e.g. [`BlindBeaconBlock`](#blindbeaconblock), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and use the domain specified for beacon block proposals.
* Builder API messages, e.g. [`builder_setFeeRecipientV1`](#builder_setFeeRecipientV1) and the response to [`builder_getHeader`](#response-2), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and the domain `DomainType('0xXXXXXXXX')` (TODO: get a proper domain).

As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).

> ## Methods

Author: Missing health endpoint point?

Reviewer: Yes! See also https://github.com/flashbots/mev-boost/issues/60 -- the only question is about naming, and which response. Did we leave the workshop with the consensus that a JSON-RPC call is fine?

If so, we could call the method `mevboost_health` or `mevboost_status` perhaps, and the response be just `true` for now? I feel like reusing the `builder` namespace isn't great, because it's literally querying the mevgeth status, not the builder status 🤔

Author: We did agree that a JSON-RPC call was fine, but did not come up with a name. It does seem like `builder` isn't a perfect fit, but I don't think it's awkward enough to need a new namespace. `builder_status` seems like an arbitrary name that can be used by both `mev-boost` and relays, should they choose, to provide general status information.

> ### `builder_setFeeRecipientV1`

Author: From Danny: maybe this should be more generic ("builder_setValidatorParams")?

Author: "builder_registerValidator"

Reviewer: we've agreed in the workshop yesterday to change it! the proposal was `builder_registerValidator`

#### Request

- method: `builder_setFeeRecipientV1`
- params:
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.

>   2. `publicKey`: `DATA`, 48 Bytes - Public key of validator.

Reviewer: Iirc Danny mentioned that the specs put all relevant fields into the message, and in this case we should add `publicKey` into it.

> ### `builder_setFeeRecipientV1`
> 
> #### Request
> 
> - method: `builder_setFeeRecipientV1`
> - params:
>   1. `message`: `object`
>       1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
>       2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
>   2. `publicKey`: `DATA`, 48 Bytes - Public key of validator.
>   3. `signature`: `DATA`, 96 Bytes - Signature over `feeRecipient` and `timestamp`.

Reviewer: ```suggestion
### `builder_registerValidator`

#### Request

- method: `builder_registerValidator`
- params:
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
      3. `pubkey`: `DATA`, 48 Bytes - Public key of validator.
  2. `signature`: `DATA`, 96 Bytes - Signature over `feeRecipient` and `timestamp`.
```

As discussed in the workshop, move `publicKey` into the message and rename to `pubkey` (as in other consensus specs)

#### Response

> - result: `null`

Reviewer: ```suggestion
- result: `true`
```

Does `true` better convey that the operation was successful?

Author: I changed this to return `"OK"` when successful.

- error: code and message set in case an exception happens while getting the payload.

#### Specification

> 1. Builder software **MUST** verify `signature` is valid under `publicKey`.
> 2. Builder software **MUST** respond to requests where `timestamp` is before the latest announcement from the validator with `-32006: Invalid timestamp`.
> 3. Builder software **MUST** store `feeRecipient` in a map keyed by `publicKey`.

Reviewer: ```suggestion
1. Builder software **MUST** verify `signature` is valid under `publicKey`.
2. Builder software **MUST** respond to requests where `timestamp` is before the latest announcement from the validator with `-32006: Invalid timestamp`.
3. Builder software **MUST** respond to requests where `timestamp` is more than 1 hour in the future `-32006: Invalid timestamp`.
4. Builder software **MUST** store `feeRecipient` in a map keyed by `publicKey`.
```

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:

>   1. `hash`: `DATA`, 32 Bytes - Hash of Execution Layer block which the validator intends to use as the parent for its proposal.

Author: Add `slot` + `pubkey` here. The public key is needed to differentiate between forks on epoch boundaries where a slashing event causes the subsequent proposer shuffling to be changed.

#### Response

> - result: `object`
>     - `message`: `object`
>         - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
>         - `value`: `DATA`, 32 Bytes - the payment in wei that will be directed to the `feeRecipient` account.
>     - `publicKey`: `DATA`, 48 Bytes - the public key associated with the builder.
>     - `signature`: `DATA`, 96 Bytes - BLS signature of the builder over `payload` and `value`.

Reviewer: ```suggestion
- result: `object`
    - `message`: `object`
        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
        - `value`: `DATA`, 32 Bytes - the payment in wei that will be directed to the `feeRecipient` account.
        - `pubkey`: `DATA`, 48 Bytes - the public key associated with the builder.
    - `signature`: `DATA`, 96 Bytes - BLS signature of the builder over `payload` and `value`.
```

As discussed in the workshop, here too `publicKey` should go into the message. And by convention it should be called `pubkey` (as in other consenus specs)

- error: code and message set in case an exception happens while getting the payload.

#### Specification
1. Builder software **SHOULD** respond immediately with the `header` that increases the `feeRecipient`'s balance by the most.
2. Builder software **MUST** return `-32001: Unknown hash` if the block identified by `hash` does not exist.
3. Builder software **MUST** return `-32002: Unknown validator` if the validator the builder expects to propose in the current slot has not been mapped to a `feeRecipient`.
4. Builder software **MAY** set the `feeRecipient` for the block to a different address than the address mapped to the validator so long as a payment equal to `value` is made to `feeRecipient`.

### `builder_getPayloadV1`

#### Request

- method: `builder_getPayloadV1`
- params:

>   1. `block`: `DATA`, arbitray length - SSZ encoded [`SignedBlindBeaconBlock`](#blindbeaconblock).

Author: Break out into JSON value.

#### Response

> - result: [`ExecutionPayloadV1`](#executionpayloadv1).

Author: It seems like this is missing a signature and public key from the relay. Without this, it doesn't seem like it's attributable if they reveal the wrong information. I don't think this is absolutely critical as withholding is still worse and not attributable, but it seems like we should add it on principle.

> - result: [`ExecutionPayloadV1`](#executionpayloadv1).

Author: @paulhauner asked if we should return the entire beacon block?

# Builder API

This document specifies the Builder API methods that the Consensus Layer uses to interact with external block builders.

## Structures

### `ExecutionPayloadV1`

Mirror of [`ExecutionPayloadV1`][execution-payload].

### `ExecutionPayloadHeaderV1`
Equivalent to `ExecutionPayloadV1`, except `transactions` is replaced with `transactionsRoot`.
- `parentHash`: `DATA`, 32 Bytes
- `feeRecipient`:  `DATA`, 20 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactionsRoot`: `DATA`, 32 Bytes

> ### `BlindBeaconBlockV1`

Reviewer 2: The beacon API uses 'blinded' rather than 'blind'.  Would be good to have consistency here.

Author: done

- `data`: `object`, [`AttestationDataV1`](#attestationdatav1)
- `signature`: `DATA`, 96 Bytes

### `DepositV1`
- `proof`: `Array`, 32 Bytes
- `data`: `object`, [`DepositDataV1`](#depositdatav1)

### `DepositDataV1`
- `pubkey`: `DATA`, 48 Bytes
- `withdrawalCredentials`: `DATA`, 32 Bytes
- `amount`: `QUANTITY`, 64 Bits
- `signature`: `DATA`, 96 Bytes

### `VoluntaryExitV1`
- `epoch`: `QUANTITY`, 64 Bits
- `ValidatorIndex`: `QUANTITY`, 64 Bits

### `SyncAggregateV1`
- `syncCommitteeBits`: `DATA`, 64 Bytes
- `syncCommitteeSignature`: `DATA`, 96 Bytes

#### Signed Containers

### `SignedBeaconBlockHeaderV1`
- `message`: `object`, [`BeaconBlockHeader`](#beaconblockheaderv1)
- `signature`: `DATA`, 96 Bytes

### `SignedVoluntaryExitV1`
- `message`: `object`, [`VoluntaryExitV1`](#voluntaryexitv1)
- `signature`: `DATA`, 96 Bytes

#### SSZ Objects

Consider the following definitions supplementary to the definitions in [`consensus-specs`][consensus-specs].

##### `ValidatorRegistrationV1`

```python
class ValidatorRegistrationV1(Container):
    feeRecipient: Bytes20
    timestamp: uint64
    pubkey: BLSPubkey
```

##### `BuilderReceiptV1`

```python

> class BuilderBidV1(Container):

Reviewer 3: shouldn't be `class BuilderReceiptV1(Container)`?

Author: Updated - merged some suggested commits, but missed a few other areas it was used.

>     header: ExecutionPayloadHeader

Reviewer 3: `execution_payload_header: ExecutionPayloadHeader` for consistency?

Author: IMO, this is unnecessarily verbose in this context.

#### Specification
1. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.
2. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 1 hour in the future with error `-32007: Invalid timestamp`.
3. Builder software **MUST** store `feeRecipient` in a map keyed by `pubkey`.
4. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:
  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.
  2. `pubkey`: `QUANTITY`, 48 Bytes - BLS public key of validator.
  3. `hash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

#### Response

- result: `object`
    - `message`: `object`
        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
        - `value`: `DATA`, 32 Bytes - Payment in wei that will be paid to the `feeRecipient` account.
        - `pubkey`: `DATA`, 48 Bytes - BLS public key associated with the builder.

>     - `signature`: `DATA`, 96 Bytes - BLS signature over [`BuilderBid1`](#builderbidv1).

Reviewer 2: Should be `BuilderBidV1`, not `BuilderBid1`.

### `CheckpointV1`
- `epoch`: `QUANTITY`, 64 Bits
- `root`: `DATA`, 32 Bytes

### `AttestationV1`
- `aggregationBits`: `DATA`, 0 to 256 Bytes
- `data`: `object`, [`AttestationDataV1`](#attestationdatav1)
- `signature`: `DATA`, 96 Bytes

### `DepositV1`
- `proof`: `Array`, 32 Bytes
- `data`: `object`, [`DepositDataV1`](#depositdatav1)

### `DepositDataV1`
- `pubkey`: `DATA`, 48 Bytes
- `withdrawalCredentials`: `DATA`, 32 Bytes
- `amount`: `QUANTITY`, 64 Bits
- `signature`: `DATA`, 96 Bytes

### `VoluntaryExitV1`
- `epoch`: `QUANTITY`, 64 Bits
- `ValidatorIndex`: `QUANTITY`, 64 Bits

### `SyncAggregateV1`
- `syncCommitteeBits`: `DATA`, 64 Bytes
- `syncCommitteeSignature`: `DATA`, 96 Bytes

#### Signed Containers

### `SignedBeaconBlockHeaderV1`
- `message`: `object`, [`BeaconBlockHeader`](#beaconblockheaderv1)
- `signature`: `DATA`, 96 Bytes

### `SignedVoluntaryExitV1`
- `message`: `object`, [`VoluntaryExitV1`](#voluntaryexitv1)
- `signature`: `DATA`, 96 Bytes

#### SSZ Objects

Consider the following definitions supplementary to the definitions in [`consensus-specs`][consensus-specs].

##### `ValidatorRegistrationV1`

```python

> class ValidatorRegistrationV1(Container):
>     feeRecipient: Bytes20
>     timestamp: uint64
>     pubkey: BLSPubkey

Reviewer: ```suggestion
class ValidatorRegistrationV1(Container):
    feeRecipient: Bytes20
    timestamp: uint64
    pubkey: BLSPubkey
    gasLimit: uint64
```

{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": -32000,
    "message": "Server error",
    "data": {
        "err": "Database corrupted"
    }
  }
}
```

## Routines

### Signing

All signature operations should follow the [standard BLS operations][bls] interface defined in `consensus-specs`.

There are two types of data to sign over in the Builder API:
* In-protocol messages, e.g. [`BlindedBeaconBlock`](#blindedbeaconblock), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and use the domain specified for beacon block proposals.
* Builder API messages, e.g. [`builder_registerValidatorV1`](#builder_registerValidatorV1) and the response to [`builder_getHeader`](#response-2), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and the domain `DomainType('0xXXXXXXXX')` (TODO: get a proper domain).

As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).

## Methods

> ### `builder_status`

Reviewer 4: clarification:

does this endpoint serve as a liveness check for the `mev_boost` server? or the upstream relays?

Reviewer: Agree, this should be clarified, at least in the spec with an explanatory sentence! From the context I had on this, it was for only querying status of the `mev_boost` server, which is also the current implementation (just checks whether it's alive).

It could be extended to the relays and return OK if at least one relay returns ok, if that would be more useful.

Reviewer: Thinking about it, I think maybe this call should return an object that can later be extended with relay liveness in a backward compatible manner.

Reviewer 4: yeah agreed, my understanding was just to ping the local `mev-boost` server to ensure it is up

can extend down the line to check for relay liveness although that seems like the kind of thing that should almost just be internal to the `mev-boost` engine so validator clients don't have to bother with it

Author: I feel like it is an implementation detail of `mev-boost` on how it should respond. I don't think this endpoint should be extended to understand the concept of multiple relays. IMO the simplicity of this comes from assuming `mev-boost` multiplexes the responses. `mev-boost` should decided what number of relays it needs an `"OK"` from to be confident it will be able to produce a block.

#### Request

- method: `builder_status`
- params: `null`

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case the builder is not operating normally.

#### Specification

1. Builder software **MUST** return `-32000: Server error` if it is unable to respond to requests. `err` **SHOULD** be set explaining the issue and `result` **MUST** be `null`.

### `builder_registerValidatorV1`

#### Request

- method: `builder_registerValidatorV1`
- params:
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
      3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.

>   3. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).

Reviewer: ```suggestion
  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).
```

>   1. `message`: `object`
>       1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
>       2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
>       3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
>   3. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).

Reviewer: ```suggestion
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
      3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
      4. `gasLimit`: `QUANTITY` - target block gas limit
  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).
```

added `gasLimit`

Equivalent to `ExecutionPayloadV1`, except `transactions` is replaced with `transactionsRoot`.
- `parentHash`: `DATA`, 32 Bytes
- `feeRecipient`:  `DATA`, 20 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactionsRoot`: `DATA`, 32 Bytes

### `BlindedBeaconBlockV1`
- `slot`: `QUANTITY`, 64 Bits
- `proposerIndex`: `QUANTITY`, 64 Bits
- `parentRoot`: `DATA`, 32 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `body`: `object`, [`BlindedBeaconBlockBodyV1`](#blindedbeaconblockbodyv1)

### `BlindedBeaconBlockBodyV1`
- `randaoReveal`: `DATA`, 96 Bytes
- `eth1Data`: `object`, [`Eth1DataV1`](#eth1datav1)
- `graffiti`: `DATA`, 32 Bytes
- `proposerSlashings`: `Array`, [`ProposerSlashingV1`](#proposerslashingv1)
- `attesterSlashings`: `Array`, [`AttesterSlashingV1`](#attesterslashingv1)
- `attestations`: `Array`, [`AttestationV1`](#attestationv1)

> - `deposits`: `Array`, [`DespositV1`](#depositv1)

Reviewer 5: There's a typo in `De**s**positV1`, right ?

> - `deposits`: `Array`, [`DespositV1`](#depositv1)

Reviewer 4: nice catch, here is a button light client can click to fix

```suggestion
- `deposits`: `Array`, [`DepositV1`](#depositv1)
```

As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).

## Methods

### `builder_status`

#### Request

- method: `builder_status`
- params: `null`

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case the builder is not operating normally.

#### Specification

1. Builder software **MUST** return `-32000: Server error` if it is unable to respond to requests. `err` **SHOULD** be set explaining the issue and `result` **MUST** be `null`.

### `builder_registerValidatorV1`

#### Request

- method: `builder_registerValidatorV1`
- params:
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `gasLimit`: `QUANTITY`, 64 bits - Target gas limit for block.
      3. `timestamp`: `QUANTITY`, 64 bits - Unix timestamp of announcement.

>       4. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.

Reviewer: Sidenote: Depending on https://github.com/ethereum/beacon-APIs/pull/206, `pubkey` might be replaced with `validatorIndex`. This can be changed in a follow-up PR when needed.

Reviewer 4: the thing w/ `validatorIndex` is that we would like the option to re-use indices in the future (we aren't anywhere near that yet but where possible designing new feature to be forwards compatible w/ this possibility)

`pubkey`s will always uniquely identify a validator, there could be a world where `index` does not

for this use case, it would always be clear who the index refers to but it is something to keep in mind in the event there are knock-on effects down the road to software in this stack

Reviewer: I think we can expect builders look up the `pubkey` by index shortly before block construction. If we want to have `pubkey` be part of this call, we might need to it to https://github.com/ethereum/beacon-APIs/pull/206 as well.

Reviewer 6: > the thing w/ `validatorIndex` is that we would like the option to re-use indices in the future (we aren't anywhere near that yet but where possible designing new feature to be forwards compatible w/ this possibility)

Good point. As long as registration request validity is restricted to `10s` or any other small amount of time this shouldn't be an issue. I admit that relying on timeout as a replay protection in this case may lead to unpredictable consequences if timeout for some reason will be removed in the future. If it isn't difficult for implementations to add `pubkey` to BN api request then we better have `pubkey`, furthermore, beacon API supports querying validators by `pubkey` which can be utilised by builders to verify that such validator exists in a validator set.

#### Specification
1. Builder software **MUST** verify `pubkey` corresponds to an active or pending validator, otherwise return error `-32002: Unknown validator`.
2. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.
3. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 10 seconds in the future with error `-32007: Invalid timestamp`.
4. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:
  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.
  2. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
  3. `parentHash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

#### Response

- result: `object`
    - `message`: `object`
        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
        - `value`: `QUANTITY`, 256 Bits - Payment in wei that will be paid to the `feeRecipient` account.

>         - `pubkey`: `DATA`, 48 Bytes - BLS public key associated with the builder.

Reviewer: Sidenote: this might actually be the pubkey of the relay, or we might want signatures by both builder and relay. Definitely need the relay signature if the validator wants to whitelist a specific relay and require correct signatures from the relay.

    - `signature`: `DATA`, 96 Bytes - BLS signature over [`BuilderBidV1`](#builderbidv1).
- error: code and message set in case an exception happens while getting the header.

> #### Specification

Reviewer 7: Should there be a restriction to disallow requests where input `slot` is in the future? 
I'm not 100% sure if that's necessary, but I wonder what would be the intended behavior if anyone calls `builder_getHeaderV1` earlier than the assigned slot. Curious to hear ppl's thoughts

Reviewer: The relay/builder should perhaps just block such a request and return an error.

3. Builder software **MUST** return `-32002: Unknown validator` if `pubkey` does not map to the validator that is expected to propose at `slot`.
4. Builder software **MUST** return `-32003: Unknown fee recipient` if the builder does not have a `feeRecipient` mapped to the validator.
5. Builder software **MAY** set the `feeRecipient` for the block to a different address than the address mapped to the validator so long as a payment equal to `value` is made to `feeRecipient`.
6. Builder software **MUST** return a header whose `gasLimit` is equal to the validator's registered `gasLimit` or as close as possible under the constraints of the consensus rules.

### `builder_getPayloadV1`

#### Request

- method: `builder_getPayloadV1`
- params:
  1. `message`: [`BlindedBeaconBlock`](#blindedbeaconblockv1).
  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`BlindedBeaconBlock`](#blindedbeaconblockv1).

#### Response

- result: [`ExecutionPayloadV1`](#executionpayloadv1).
- error: code and message set in case an exception happens while proposing the payload.

#### Specification

> 1. Builder software **MUST** be able to un-blind the [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1) in `message`, otherwise the return `-32004: Unknown block`.

Reviewer 5: ```suggestion
1. Builder software **MUST** be able to un-blind the [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1) in `message`, otherwise return the error `-32004: Unknown block`.
```

* Builder API messages, e.g. [`builder_registerValidatorV1`](#builder_registerValidatorV1) and the response to [`builder_getHeader`](#response-2), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and the domain `DomainType('0xXXXXXXXX')` (TODO: get a proper domain).

As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).

## Methods

### `builder_status`

#### Request

- method: `builder_status`
- params: `null`

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case the builder is not operating normally.

#### Specification

1. Builder software **MUST** return `-32000: Server error` if it is unable to respond to requests. `err` **SHOULD** be set explaining the issue and `result` **MUST** be `null`.

### `builder_registerValidatorV1`

#### Request

- method: `builder_registerValidatorV1`
- params:

>   1. `message`: `object`
>       1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
>       2. `gasTarget`: QUANTITY, 64 bits - Target for block `gasTarget`.
>       2. `timestamp`: `QUANTITY`, 64 bits - Unix timestamp of announcement.
>       3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
>   3. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).

Reviewer: ```suggestion
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `gasTarget`: QUANTITY, 64 bits - Target for block `gasTarget`.
      3. `timestamp`: `QUANTITY`, 64 bits - Unix timestamp of announcement.
      4. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).
```

fixed the numbering

#### Response

> - result: `enum`, `"OK" | null`

Reviewer 4: what happens if we register with multiple relays/builders and some fail while others succeed?

seems like this should return an array of responses

Author: I think this is the choice of `mev-boost` whether to succeed or fail if not all of it's requests succeed. I think it might make sense to extend the error object so it's easier to figure out which relay caused the error, but not sure I'd go as far as returning an array of all the relay responses.

Reviewer: Agree with Matt here. Also, mev-boost continuously resends the `registerValidator` call to the relays/builders..

- error: code and message set in case an exception happens while registering the validator.

#### Specification

> 1. Builder software **MUST** verify `pubkey` corresponds to an active or pending validator, otherwise return error `-32003: Unknown validator`.

Reviewer 4: ```suggestion
1. Builder software **MUST** verify `pubkey` corresponds to an active or pending validator, otherwise return error `-32002: Unknown validator`.
```

2. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.

> 3. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 1 hour in the future with error `-32007: Invalid timestamp`.

Reviewer: ```suggestion
3. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 10 seconds in the future with error `-32007: Invalid timestamp`.
```

> 4. Builder software **MUST** store `feeRecipient` in a map keyed by `pubkey`.

Reviewer: Point 4 seems like an implementation detail, should this even be part of the spec?

Reviewer 4: +1

Author: Updated

5. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:
  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.

>   2. `pubkey`: `QUANTITY`, 48 Bytes - BLS public key of validator.

Reviewer 3: ```suggestion
  2. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
```

  3. `parentHash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

#### Response

- result: `object`
    - `message`: `object`
        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).

>         - `value`: `DATA`, 32 Bytes - Payment in wei that will be paid to the `feeRecipient` account.

Reviewer 3: ```suggestion
        - `value`: `QUANTITY`, 256 Bits - Payment in wei that will be paid to the `feeRecipient` account.
```

# Builder API

This document specifies the Builder API methods that the Consensus Layer uses to interact with external block builders.

> 

Reviewer 4: it may be worth adding a note that "builders" in this context are *trusted*

so that we can drop a lot of validations you would otherwise want to do

e.g. we can assume the execution header the validator gets back in `builder_getHeaderV1` is filled out in a way that reflects the validator's wishes so do not need to specify that the header is validated before signing over it bc builders in this iteration are trusted to do their job well (with a corresponding hit to reputation when they fail)

if we move to more decentralized alternatives it will likely become more important to validate more of the data we receive

Author: I agree, but I think it is important to still have an adversarial mindset when interacting with builders. If there areas of significant trust, we should work to reduce it. IMO the validator should validate the header they receive to the best of their ability before signing.

```mermaid
sequenceDiagram
    participant consensus

>     participant mev_boost

Reviewer 7: Should we explicitly mention `mev_boost` here? I wonder if the mermaid diagram is even necessary, given the other specs don't have it

Reviewer 4: could call this "builder software" or something like this

I do think it is worth removing mention to `mev_boost` unless we want to enshrine that name

Author: Removed the diagram.

- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactionsRoot`: `DATA`, 32 Bytes

### `BlindBeaconBlockV1`
- `slot`: `QUANTITY`, 64 Bits
- `proposerIndex`: `QUANTITY`, 64 Bits
- `parentRoot`: `DATA`, 32 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `body`: `object`, [`BlindBeaconBlockBodyV1`](#blindbeaconblockbodyv1)

### `BlindBeaconBlockBodyV1`
- `randaoReveal`: `DATA`, 96 Bytes
- `eth1Data`: `object`, [`Eth1DataV1`](#eth1datav1)
- `graffiti`: `DATA`, 32 Bytes
- `proposerSlashings`: `Array`, [`ProposerSlashingV1`](#proposerslashingv1)
- `attesterSlashings`: `Array`, [`AttesterSlashingV1`](#attesterslashingv1)
- `attestations`: `Array`, [`AttestationV1`](#attestationv1)
- `deposits`: `Array`, [`DespositV1`](#depositv1)
- `voluntaryExits`: `Array`, [`SignedVoluntaryExitV1`](#signedvoluntaryexitv1)
- `syncAggregate`: `object`, [`SyncAggregateV1`](#syncaggregatev1)

> - `executionPayload`: `object`, [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1)

Reviewer 4: ```suggestion
- `executionPayloadHeader`: `object`, [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1)
```

- `data`: `object`, [`AttestationDataV1`](#attestationdatav1)
- `signature`: `DATA`, 96 Bytes

### `DepositV1`
- `proof`: `Array`, 32 Bytes
- `data`: `object`, [`DepositDataV1`](#depositdatav1)

### `DepositDataV1`
- `pubkey`: `DATA`, 48 Bytes
- `withdrawalCredentials`: `DATA`, 32 Bytes
- `amount`: `QUANTITY`, 64 Bits
- `signature`: `DATA`, 96 Bytes

### `VoluntaryExitV1`
- `epoch`: `QUANTITY`, 64 Bits
- `ValidatorIndex`: `QUANTITY`, 64 Bits

### `SyncAggregateV1`
- `syncCommitteeBits`: `DATA`, 64 Bytes
- `syncCommitteeSignature`: `DATA`, 96 Bytes

#### Signed Containers

### `SignedBeaconBlockHeaderV1`
- `message`: `object`, [`BeaconBlockHeader`](#beaconblockheaderv1)
- `signature`: `DATA`, 96 Bytes

### `SignedVoluntaryExitV1`
- `message`: `object`, [`VoluntaryExitV1`](#voluntaryexitv1)
- `signature`: `DATA`, 96 Bytes

#### SSZ Objects

Consider the following definitions supplementary to the definitions in [`consensus-specs`][consensus-specs].

##### `ValidatorRegistrationV1`

```python
class ValidatorRegistrationV1(Container):
    feeRecipient: Bytes20
    timestamp: uint64
    pubkey: BLSPubkey
```

##### `BuilderReceiptV1`

```python

> class BuilderReceiptV1(Container):

Reviewer 4: name change?

```suggestion
class BuilderBidV1(Container):
```

I think "bids" is more appropriate here than "receipts" (which implies the output of a finished process) and this is in the middle of what could become a very dynamic process

this terminology is also forwards compatible with any PBS stuff we do down the line

>     payload: ExecutionPayloadHeader

Reviewer 4: ```suggestion
    execution_payload_header: ExecutionPayloadHeader
```

>     payload: ExecutionPayloadHeader

Reviewer 6: ```suggestion
    header: ExecutionPayloadHeader
```

    value: uint256
    pubkey: BLSPubkey
```

###### `BlindBeaconBlock`

```python
class BlindBeaconBlock(Container):
    slot: Slot
    proposer_index: ValidatorIndex
    parent_root: Root
    state_root: Root
    body: BlindBeaconBlockBody
```

###### `BlindBeaconBlockBody`

```python
class BlindBeaconBlockBody(Container):
    randao_reveal: BLSSignature
    eth1_data: Eth1Data
    graffiti: Bytes32
    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
    attestations: List[Attestation, MAX_ATTESTATIONS]
    deposits: List[Deposit, MAX_DEPOSITS]
    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
    sync_aggregate: SyncAggregate

>     execution_payload: ExecutionPayloadHeader

Reviewer 4: ```suggestion
    execution_payload_header: ExecutionPayloadHeader
```

## Methods

### `builder_status`

#### Request

- method: `builder_status`
- params: `null`

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case the builder is not operating normally.

#### Specification

1. Builder software **MUST** return `-32000: Server error` if it is unable to respond to requests. `err` **SHOULD** be set explaining the issue and `result` **MUST** be `null`.

### `builder_registerValidatorV1`

#### Request

- method: `builder_registerValidatorV1`
- params:
  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
      3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
  3. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case an exception happens while registering the validator.

#### Specification

> 1. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.

Reviewer 6: Shouldn't belonging of `pubkey` to existing validator set be checked? One could be signing the same message with thousands of different pubkeys and DoS builder/relay.

Author: done

> 2. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 1 hour in the future with error `-32007: Invalid timestamp`.

Reviewer 4: any rationale on why "1 hour in the future"?

the only reason to accept future honest requests is clock skew and I think if someone's clock is off by ~1 hr they will have bigger problems.

I'd think to tighten this to 60 seconds or even less

Author: updated to 10 sec

3. Builder software **MUST** store `feeRecipient` in a map keyed by `pubkey`.
4. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:
  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.
  2. `pubkey`: `QUANTITY`, 48 Bytes - BLS public key of validator.

>   3. `hash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

Reviewer 3: `parent_hash` or `parentHash`? It seems like we are serializing to json in camelCase

>   3. `hash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

Reviewer 4: ah yes, sorry `camelCase`!

#### Specification
1. Builder software **SHOULD** respond immediately with the `header` that increases the `feeRecipient`'s balance by the most.
2. Builder software **MUST** return `-32001: Unknown hash` if the block identified by `hash` is not known.
3. Builder software **MUST** return `-32002: Unknown validator` if `pubkey` does not map to the validator that is expected to propose at `slot`.
4. Builder software **MUST** return `-32003: Unknown fee recipient` if the builder does not have a `feeRecipient` mapped to the validator.
5. Builder software **MAY** set the `feeRecipient` for the block to a different address than the address mapped to the validator so long as a payment equal to `value` is made to `feeRecipient`.

### `builder_getPayloadV1`

#### Request

- method: `builder_getPayloadV1`
- params:
  1. `message`: [`BlindBeaconBlock`](#blindbeaconblockv1).
  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`BlindBeaconBlock`](#blindbeaconblockv1).

#### Response

- result: [`ExecutionPayloadV1`](#executionpayloadv1).
- error: code and message set in case an exception happens while proposing the payload.

#### Specification

> 1. Builder software **MUST** verify that the beacon block's exeuction payload is a matching [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1) provided from [`builder_getHeaderV1`](#buildergetheaderv1), otherwise the return `-32004: Unknown block`.

Reviewer 6: For this verification builder software will have to keep recent `build_getHeaderV1` call parameters. Wouldn't it be enough to match *signed* `executionPayloadHeader` provided by CL on the set of headers known to builder software? If there is a match the only thing matters is that a validator is signed off on certain header even though it's not the one that was sent to the validator in the previous roundtrip.

Author: You're right, the current text is overly prescriptive.

> 1. Builder software **MUST** verify that the beacon block's exeuction payload is a matching [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1) provided from [`builder_getHeaderV1`](#buildergetheaderv1), otherwise the return `-32004: Unknown block`.

Reviewer 4: ```suggestion
1. Builder software **MUST** verify that the beacon block's execution payload is a matching [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1) provided from [`builder_getHeaderV1`](#buildergetheaderv1), otherwise the return `-32004: Unknown block`.
```

> 2. Builder software **MUST** verify that `signature` is a BLS signature over `block` using [`verify_block_signature`][verify-block-signature] from the validator that is expected to propose in the slot. If the signature is determined to be invalid or from a different validator than expected, the builder **MUST** return `-32005: Invalid signature`.

Reviewer 6: Note, that signature must be verified with respect to a proposer index computed from a pre-state of given `BlindBeaconBlock` as theoretically there could be two competing chains with different shuffling therefore different proposer indexes for the same slot.

Author: noted

    mev_boost-->>consensus: builder_getPayload response
```

## Structures

### `ExecutionPayloadV1`

Mirror of [`ExecutionPayloadV1`][execution-payload].

### `ExecutionPayloadHeaderV1`
Equivalent to `ExecutionPayloadV1`, except `transactions` is replaced with `transactionsRoot`.
- `parentHash`: `DATA`, 32 Bytes
- `feeRecipient`:  `DATA`, 20 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactionsRoot`: `DATA`, 32 Bytes

### `BlindBeaconBlockV1`
- `slot`: `QUANTITY`, 64 Bits
- `proposerIndex`: `QUANTITY`, 64 Bits
- `parentRoot`: `DATA`, 32 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `body`: `object`, [`BlindBeaconBlockBodyV1`](#blindbeaconblockbodyv1)

### `BlindBeaconBlockBodyV1`
- `randaoReveal`: `DATA`, 96 Bytes
- `eth1Data`: `object`, [`Eth1DataV1`](#eth1datav1)
- `graffiti`: `DATA`, 32 Bytes

> - `proposerSlashings`: `object`, [`ProposerSlashingV1`](#proposerslashingv1)
> - `attesterSlashings`: `object`, [`AttesterSlashingV1`](#attesterslashingv1)

Reviewer 7: I think slashings are also arrays given they are a `List`?

```suggestion
- `proposerSlashings`: `Array`, [`ProposerSlashingV1`](#proposerslashingv1)
- `attesterSlashings`: `Array`, [`AttesterSlashingV1`](#attesterslashingv1)
```

### `AttesterSlashingV1`
- `attestation1`: `object`, [`IndexedAttestationV1`](#indexedattestationv1)
- `attestation2`: `object`, [`IndexedAttestationV1`](#indexedattestationv1)

### `IndexedAttestationV1`
- `attestingIndices`: `Array`, `QUANTITY`, 64 Bits
- `data`: `object`, [`AttestationDataV1`](#attestationdatav1)
- `signature`: `DATA`, 96 Bytes

### `AttestationDataV1`
- `slot`: `QUANTITY`, 64 Bits
- `index`: `QUANTITY`, 64 Bits
- `beaconBlockRoot`, `DATA`, 32 Bytes
- `source`: `object`, [`CheckpointV1`](#checkpointv1)
- `target`: `object`, [`CheckpointV1`](#checkpointv1)

### `CheckpointV1`
- `epoch`: `QUANTITY`, 64 Bits
- `root`: `DATA`, 32 Bytes

### `AttestationV1`
- `aggregationBits`: `DATA`, 0 to 256 Bytes
- `data`: `object`, [`AttestationDataV1`](#attestationdatav1)
- `signature`: `DATA`, 96 Bytes

### `DepositV1`
- `proof`: `Array`, 32 Bytes
- `data`: `object`, [`DepositDataV1`](#depositdatav1)

### `DepositDataV1`
- `pubkey`: `DATA`, 48 Bytes
- `withdrawalCredentials`: `DATA`, 32 Bytes
- `amount`: `QUANTITY`, 64 Bits
- `signature`: `DATA`, 96 Bytes

### `VoluntaryExitV1`
- `epoch`: `QUANTITY`, 64 Bits
- `ValidatorIndex`: `QUANTITY`, 64 Bits

### `SyncAggregateV1`

> - `syncCommitteeBits`: `DATA`, 0 to 64 Bytes

Reviewer 7: sync committee bits are fixed to 512 bits (`Bitvector[SYNC_COMMITTEE_SIZE]`). Shouldn't it just be? 👇🏼 
```suggestion
- `syncCommitteeBits`: `DATA`, 64 Bytes
```

Author: Ah good catch, thanks.

- `syncCommitteeSignature`: `DATA`, 96 Bytes

#### Signed Containers

### `SignedBeaconBlockHeaderV1`
- `message`: `object`, [`BeaconBlockHeader`](#beaconblockheaderv1)
- `signature`: `DATA`, 96 Bytes

### `SignedVoluntaryExitV1`
- `message`: `object`, [`VoluntaryExitV1`](#voluntaryexitv1)
- `signature`: `DATA`, 96 Bytes

#### SSZ Objects

Consider the following definitions supplementary to the definitions in [`consensus-specs`][consensus-specs].

##### `ValidatorRegistrationV1`

```python
class ValidatorRegistrationV1(Container):
    feeRecipient: Bytes20
    timestamp: uint64
    pubkey: BLSPubkey
```

##### `BuilderReceiptV1`

```python
class BuilderReceiptV1(Container):
    payload: ExecutionPayloadHeader
    value: uint256
    pubkey: BLSPubkey
```

> ##### `SignedBlindBeaconBlock`
> 
> ###### `SignedBlindBeaconBlock`
> 
> ```python
> class SignedBlindBeaconBlock(Container):
>     message: BlindBeaconBlock
>     signature: BLSSignature
> ```
> 

Reviewer: ```suggestion
```

Not needed anymore

  1. `message`: `object`
      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
      2. `timestamp`: `QUANTITY`, uint64 - Unix timestamp of announcement.
      3. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
  3. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).

#### Response

- result: `enum`, `"OK" | null`
- error: code and message set in case an exception happens while registering the validator.

#### Specification
1. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.
2. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 1 hour in the future with error `-32007: Invalid timestamp`.
3. Builder software **MUST** store `feeRecipient` in a map keyed by `pubkey`.
4. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.

### `builder_getHeaderV1`

#### Request

- method: `builder_getHeaderV1`
- params:
  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.

>   2. `pubkey`: `QUANTITY`, 64 Bits - BLS public key of proposer.

Reviewer: ```suggestion
  2. `pubkey`: `QUANTITY`, 48 Bytes - BLS public key of validator.
```

  3. `hash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.

#### Response

- result: `object`
    - `message`: `object`
        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
        - `value`: `DATA`, 32 Bytes - Payment in wei that will be paid to the `feeRecipient` account.

>         - `pubkey`: `DATA`, 48 Bytes - BLS public key associated with the builder.

Reviewer 7: Isn't this just `BuilderReceiptV1`? if yes, couldn't we just use it here and replace the 3 lines?

Author: @terencechain `BuilderReceiptV1` is defined as an SSZ object. This is just the JSON-RPC definition of the same.

Reviewer 7: Got it. Thanks! I always get mixed up btw the two...

Author: Same tbh
